provider/pkg/kinds/util.go

```diff
@@ -0,0 +1,45 @@
+package kinds
+
+import (
+	"strings"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+)
+
+func suffixSearch(s string, suffixes []string) bool {
+	for _, suffix := range suffixes {
+		if strings.HasSuffix(s, suffix) {
+			return true
+		}
+	}
+	return false
+}
+
+// IsPatchURN returns true if the URN is for a Patch resource.
+func IsPatchURN(urn resource.URN) bool {
+	urnS := urn.QualifiedType().String()
+
+	// Do a simple O(1) lookup in a set of known patch types.
+	if PatchQualifiedTypes.Has(urnS) {
+		return true
+	}
+
+	// Do a suffix search in case the resource is a component resource containing
+	// a patch resource. Eg. a component resource patch resource could have the following URN:
+	// my:component:Resource$kubernetes:apps/v1:DaemonSetPatch.
+	return suffixSearch(urnS, PatchQualifiedTypes.SortedValues())
```

Comment:
Would `urn.Type()` work here instead?

IIRC `QualifiedType` gives you `parent$child` but `Type` gives you `child`, and it seems like we only want to look up `PatchQualifiedTypes.Has(child)`.

Reply:
Oh nice! Yes, that works too. Wasn't familiar with all the methods of Resource.URN, but yes, `Type` just gives the resource type, whereas `QualifiedType` also returns the parent if there is one. Updated.